### ISSUE TITLE
Triton Backend

### ISSUE THREAD

**Issue by dbasbabasi** [SthPhoenix/InsightFace-REST#60]
Hi, It works on trt backend. I am trying to run it on the triton backend. I changed docker parameter in the deploy_trt file. It fails on warmup on triton backend. Do I need change another conf?

**Comment by SthPhoenix** [comment]
Hi! Triton backend should work, but for now it's up to you to run separate Triton Server container and provide it's url to `deploy_trt.sh` config.

Also, currently there is known issue with inference of SCRFD model with Triton backend - Triton provides outputs as non writable numpy array, but new optimized SCRFD post processing modifies net output arrays in place to avoid excessive creation of numpy arrays. For now it can be fixed by replacing lines 332-334 of `scrfd.py` with:

```
score_blob = np.copy(net_outs[idx][0])
bbox_blob =  np.copy(net_outs[idx + self.fmc][0])
kpss_blob =  np.copy(net_outs[idx + self.fmc * 2][0])
```

**Comment by dbasbabasi** [comment]
![image](https://user-images.githubusercontent.com/13520065/142484902-6a868731-69c8-4922-b26e-1bf3941ed1ba.png)

Thank you for your quick response.

Actually I have some experience with Triton but there is a problem with getting the metadata during load the model and docker container is stopping automatically. I tried to debug it. But I couldn't fix it. I used following models and conf.   

max_size=640,640
det_model=retinaface_r50_v1
rec_model=arcface_r100_v1

Docker logs:
![image](https://user-images.githubusercontent.com/13520065/142485528-d4aa3d81-4c81-43a5-bfbf-27ca9c8891dd.png)

**Comment by SthPhoenix** [comment]
Have you changed localhost to actual triton server IP:grpc port?
In docker localhost is container itself, not the host machine.

**Comment by dbasbabasi** [comment]
[deploy_trt.txt](https://github.com/SthPhoenix/InsightFace-REST/files/7565687/deploy_trt.txt)

Here is my deploy_trt file. Yeah I tried with host ip and also tried with localhost and open the port on docker run comment.

**Comment by SthPhoenix** [comment]
You shouldn't bind triton ports inside insightface-rest container, it should cause exceptions when starting triton server, or IFR container

**Comment by dbasbabasi** [comment]
Yeah I got it. I deleted ports, run inference docker after that run the deploy_trt. It looks detection model uploaded and I can see the model output list during the load but I got another error for Arcface. I am checking it. Thank you so much for your help.

![image](https://user-images.githubusercontent.com/13520065/142495253-c7dd6f1c-b686-4090-bb97-3c876f8d601c.png)

**Comment by SthPhoenix** [comment]
IFR is using shared GPU memory to communicate with triton server, it may not work if triton is on different host.

**Comment by dbasbabasi** [comment]
Yeah it works on the same machine. I could send a face detection request to Triton. But when I tried to load face rec model, it is returning Cuda shared memory error.

Also I needed to change face detection request dimension for fixing it.

**Comment by SthPhoenix** [comment]
I have just checked - everything seems to be working using fix from https://github.com/SthPhoenix/InsightFace-REST/issues/60#issuecomment-972911459
I have followed these steps:
1. Run `deploy_trt.sh` setting `rec_batch_size` = 32 and `det_batch_size` = 10
2. Wait until trt engines are built
3. Stop IRF container
4. Copy engines to Triton server models folder under following paths: `{triton_models}/scrfd_10g_gnkps/1/model.plan`, `{triton_models}/glintr100/1/model.plan`
5. Run Triton server, ensure it actually have started.
6. Edit `deploy_trt.sh` changing `det_batch_size` to 1 and `INFERENCE_BACKEND` to `triton` and providing valid `triton_uri` (your host machine local IP address)
7. Run `deploy_trt.sh` again.
8. Now you IFR container should be using Triton inference server.

Though you should provide valid model  configs to get use of dynamic batching.

**Comment by dbasbabasi** [comment]
Thank you so much I used onnx model for triton. It works right now for retinaface and arcface. Do you have a plan adding to age gender for triton?

**Comment by SthPhoenix** [comment]
Gender/age model is now temporarily not supported, since g/a model requires different face crop preprocessing than current glintr100 recognition models.

**Comment by dbasbabasi** [comment]
I used retinaface resnet model for face detection. I will try to run g/a model. Thank you so much for your help. If you have a recommendation for g/a, I will be really glad, otherwise I will close this issue.

**Comment by SthPhoenix** [comment]
You could implement it, but you'll have to make copies of face crops numpy arrays at recognition step, otherwise g/a estimations will be totally wrong, due to different preprocessing required for recognition and g/a estimation. 
Copying numpy arrays will hit overall performance, though I haven't tested it yet.

**Comment by dbasbabasi** [comment]
Thank you I used my own model for that as onnx. And write new client for this models. The result looks good. Your repo is awesome. Thank you so much for your help!

**Comment by SthPhoenix** [comment]
Nice to hear that! 
Have you used publicly available model for ga or have you trained your own?

**Comment by dbasbabasi** [comment]
I used my own trained models. I converted them to the onnx and write a new client for age/gender, emotion and mask detection. After the face crop, passed the cropped face to the inference. I see retina face had pretrained mask model but it looks unavailable right now.

**Comment by SthPhoenix** [comment]
> I used my own trained models. I converted them to the onnx and write a new client for age/gender, emotion and mask detection. After the face crop, passed the cropped face to the inference. I see retina face had pretrained mask model but it looks unavailable right now.

Sorry for late reply, finally got some free time )

You have separate models for GA, emotion and mask detection working on 112x112 face crops? 
That's interesting since all pretrained models for this tasks I have seen were expecting different input shape. 
Could you point out where I could find training code or models if you have used public repos?

**Comment by dbasbabasi** [comment]
Hey Yeah GA model is separated. It is not public repo I can't share it. Our models are works with retinaface. I have no idea about public GA and mask models
